File: 23.合并K个升序链表/23.合并K个升序链表.cpp
/*
 * @Descroption: LeetCode 23. 合并K个升序链表
 * @Author: EmoryHuang
 * @Date: 2021-05-27 21:05:17
 * @解题思路:
 * 方法一：顺序合并
 * 在合并两个升序链表的基础上，遍历k个链表，依次合并
 *
 * 方法二：分治
 * 用分治的方法进行合并
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
   public:
    // 合并两个升序链表
    ListNode* mergeTwoList(ListNode* lista, ListNode* listb) {
        if (!lista) return listb;
        if (!listb) return lista;
        if (lista->val > listb->val) {
            listb->next = mergeTwoList(lista, listb->next);
            return listb;
        } else {
            lista->next = mergeTwoList(lista->next, listb);
            return lista;
        }
    }
    // 方法一：顺序合并
    // ListNode* mergeKLists(vector<ListNode*>& lists) {
    //     ListNode* ans = nullptr;
    //     for (int i = 0; i < lists.size(); i++) ans = mergeTwoList(ans, lists[i]);
    //     return ans;
    // }

    // 方法二：分治
    ListNode* merge(vector<ListNode*>& lists, int l, int r) {
        if (l == r) return lists[l];
        int mid = l + (r - l) / 2;
        ListNode* l1 = merge(lists, l, mid);
        ListNode* l2 = merge(lists, mid + 1, r);
        return mergeTwoList(l1, l2);
    }
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.size() == 0) return nullptr;
        return merge(lists, 0, lists.size() - 1);
    }
};

File: Solution/23.合并K个升序链表/23.合并K个升序链表.java
/*
 * @Description: LeetCode 148. 排序链表
 * @Author: EmoryHuang
 * @Date: 2023-08-09 10:54:30
 * @解题思路:
 * 方法一：分治
 * 方法二：优先队列
 */

import java.util.*;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) {
            return null;
        }
        // 方法一：分治
        // return merge(lists, 0, lists.length - 1);

        // 方法二：优先队列
        PriorityQueue<ListNode> queue = new PriorityQueue<ListNode>(new Comparator<ListNode>() {
            @Override
            public int compare(ListNode o1, ListNode o2) {
                if (o1.val < o2.val) {
                    return -1;
                } else if (o1.val == o2.val) {
                    return 0;
                } else {
                    return 1;
                }
            }
        });
        ListNode dummy = new ListNode(0);
        ListNode cur = dummy;
        for (ListNode node : lists) {
            if (node != null) {
                queue.add(node);
            }
        }
        while (!queue.isEmpty()) {
            ListNode top = queue.poll();
            cur.next = top;
            if (top.next != null) {
                queue.add(top.next);
            }
            cur = cur.next;
        }
        return dummy.next;
    }

    ListNode merge(ListNode[] lists, int left, int right) {
        if (left == right) {
            return lists[left];
        }
        int mid = left + (right - left) / 2;
        ListNode l = merge(lists, left, mid);
        ListNode r = merge(lists, mid + 1, right);
        return mergeTwoList(l, r);
    }

    ListNode mergeTwoList(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(0);
        ListNode cur = dummy;
        while (l1 != null && l2 != null) {
            if (l1.val > l2.val) {
                cur.next = l2;
                l2 = l2.next;
            } else {
                cur.next = l1;
                l1 = l1.next;
            }
            cur = cur.next;
        }
        cur.next = l1 == null ? l2 : l1;
        return dummy.next;
    }
}